packages/core/data-transfer/src/engine/validation/provider.ts

```diff
@@ -0,0 +1,27 @@
+import { capitalize } from 'lodash/fp';
+
+import type { IDestinationProvider, ISourceProvider, ProviderType } from '../../../types';
+import { TransferEngineValidationError } from '../errors';
+
+const reject = (reason: string): never => {
+  throw new TransferEngineValidationError(`Invalid provider supplied. ${reason}`);
+};
+
+const validateProvider = <T extends ProviderType>(
+  type: ProviderType,
+  provider?: ([T] extends ['source'] ? ISourceProvider : IDestinationProvider) | null
+) => {
+  if (!provider) {
+    return reject(
+      `Expected an instance of "${capitalize(type)}Provider", but got "${typeof provider}" instead.`
```

Comment:
Same with these, it would be great to have an error object with more details separated out instead of a string